File: rust_vs_c++/interface/interface.cpp
// Hugo BERANGER - M2 MIAGE IA

#include <iostream>

using namespace std;

class Forme {
    public:
        Forme(){};
        int aire();
};

class Rectangle : public Forme {
        int largeur, longueur;
    public:
        Rectangle (int x, int y) {
            largeur = x;
            longueur = y;
        }
        int aire () {
            return largeur*longueur;
        }
};

class Carre : public Forme {
        int cote;
    public:
        Carre (int x) {
            cote = x;
        }
        int aire () {
            return cote*cote;
        }
};

int main () {
    Rectangle rectangleObjet(3,4);
    cout << "c++  - rectangle aire: " << rectangleObjet.aire() << endl;
    Carre carreObjet(3);
    cout << "c++  - carre aire: " << carreObjet.aire() << endl;
    return 0;
}

File: rust_vs_c++/interface/interface.rs
// Hugo BERANGER - M2 MIAGE IA

trait Forme {
    fn aire(&self) -> i32;
}

struct Rectangle {
    largeur: i32,
    longueur: i32,
}

struct Carre {
    cote: i32,
}

impl Forme for Rectangle {
        fn aire(&self) -> i32 {
            self.largeur * self.longueur
        }
}

impl Forme for Carre {
    fn aire(&self) -> i32 {
        self.cote * self.cote
    }
}

fn main() {
    let rectangle_objet = Rectangle {
        largeur: 3,
        longueur: 4,
    };
    println!("rust - rectangle aire: {}", rectangle_objet.aire());

    let carre_objet = Carre {
        cote:3,
    };
    println!("rust - carre aire: {}", carre_objet.aire());
}